src/audio_midiout.cpp

```diff
@@ -0,0 +1,387 @@
+
+	if (midiout->file_buffer_pos >= midiout->file_buffer.size()) {
+		return EOF;
+	}
+
+	return midiout->file_buffer[midiout->file_buffer_pos++];
+}
+
+bool GenericMidiOut::Open(Filesystem_Stream::InputStream stream) {
+	seq->clear();
+	file_buffer = Utils::ReadStream(stream);
+
+	if (!seq->load(this, read_func)) {
+		error_msg = "Midi: Error reading file";
+		return false;
+	}
+	seq->rewind();
+	mtime = seq->get_start_skipping_silence();
+
+	tempo.emplace_back(this, midi_default_tempo);
+
+	return true;
+}
+
+void GenericMidiOut::SetFade(int begin, int end, int duration) {
+	fade_steps = 0;
+	last_fade_mtime = 0.0f;
+
+	if (duration <= 0.0) {
+		SetVolume(end);
+		return;
+	}
+
+	if (begin == end) {
+		SetVolume(end);
+		return;
+	}
+
+	volume = begin / 100.0f;
+	fade_end = end / 100.0f;
+	fade_steps = duration / 100;
+	delta_step = (fade_end - volume) / fade_steps;
+}
+
+void GenericMidiOut::SetVolume(int new_volume) {
+	// cancel any pending fades
+	fade_steps = 0;
+
+	volume = new_volume / 100.0f;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+		device->SendMidiMessage(msg);
+	}
+}
+
+int GenericMidiOut::GetVolume()const {
+	if (fade_steps > 0) {
+		return static_cast<int>(fade_end * 100);
+	}
+	return static_cast<int>(volume * 100);
+}
+
+void GenericMidiOut::Pause() {
+	paused = true;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, 0);
+		device->SendMidiMessage(msg);
+	}
+}
+
+void GenericMidiOut::Resume() {
+	paused = false;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+		device->SendMidiMessage(msg);
+	}
+}
+
+void GenericMidiOut::Rewind() {
+	seq->rewind();
+}
+
+int GenericMidiOut::GetLoopCount() const {
+	return loop_count;
+}
+
+bool GenericMidiOut::IsFinished() const {
+	return seq->is_at_end();
+}
+
+bool GenericMidiOut::SetPitch(int pitch) {
+	tempo_multiplier = pitch / 100.0f;
+	return true;
+}
+
+int GenericMidiOut::GetPitch() const {
+	return static_cast<int>(tempo_multiplier * 100);
+}
+
+int GenericMidiOut::GetTicks() const {
+	assert(!tempo.empty());
+
+	return tempo.back().GetTicks(mtime);
+}
+
+void GenericMidiOut::Update(long long delta) {
+	if (paused) {
+		return;
+	}
+	if (fade_steps >= 0 && mtime - last_fade_mtime > 0.1f) {
+		volume = std::max(0.0f, std::min(1.0f, volume + delta_step));
+		for (int i = 0; i < 16; i++) {
+			uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+			device->SendMidiMessage(msg);
+		}
+		last_fade_mtime = mtime;
+		fade_steps -= 1;
+	}
+
+	seq->play(mtime, this);
+	mtime = mtime + ((delta / 1000000.0f) * tempo_multiplier);
+
+	if (IsFinished() && looping) {
+		mtime = seq->rewind_to_loop();
+		reset_tempos_after_loop();
+	}
+}
+
+bool GenericMidiOut::Seek(std::streamoff offset, std::ios_base::seekdir origin) {
+	if (offset == 0 && origin == std::ios_base::beg) {
```

Comment:
What if I seek to the beginning correctly using `cur` or `end`?

Reply:
I'm not actually convinced this function is ever used or called. Visual Studio can't find any references to it, and when I comment it all out it still builds. Might just remove this entirely

Reply:
``Seek`` is currently not really supported by any ``AudioDecoder``. This is more about future proof, just wasn't needed by now. The only case where a Seek happens right now is in ``Rewind()`` when the end of the file is reached (for BGM looping).